crates/pallet-feeds/src/feed_processor.rs

```diff
@@ -15,17 +15,82 @@
 
+use crate::CallObject;
+use codec::{Compact, Decode};
 use sp_runtime::{DispatchError, DispatchResult};
 use sp_std::{vec, vec::Vec};
-use subspace_core_primitives::{crypto, Sha256Hash};
+use subspace_core_primitives::Sha256Hash;
 
-/// Object mapping that points to an object in a block
+/// Holds the offset to some portion of data within/or the object
 #[derive(Debug)]
-pub struct FeedObjectMapping {
-    /// Key scoped to the feed
-    pub key: Sha256Hash,
-    /// Offset of the data within object
-    pub offset: u32,
+pub enum FeedObjectMapping {
+    /// Maps the entire object treating it as content
+    /// If key is provided, it is namespaced to feed to avoid collisions
+    Object { key: Option<Vec<u8>> },
+    /// Maps some data within the object at the specific offset.
+    /// Data must be a scale encoded `Vec<u8>`
+    /// If key is provided, it namespaced to feed to avoid collisions
+    Content { key: Option<Vec<u8>>, offset: u32 },
+}
+
+impl FeedObjectMapping {
+    pub(crate) fn try_into_call_object<KNS, CO, CH>(
+        self,
+        object: &[u8],
+        key_namespace: KNS,
+        call_offset: CO,
+        content_hasher: CH,
+    ) -> Option<CallObject>
+    where
+        KNS: Fn(&[u8]) -> Sha256Hash,
+        CO: Fn(u32, bool) -> u32,
+        CH: Fn(&[u8]) -> Sha256Hash,
+    {
+        match self {
+            FeedObjectMapping::Object { key } => Some(CallObject {
+                key: key
+                    .map(|key| key_namespace(key.as_slice()))
+                    .unwrap_or_else(|| content_hasher(object)),
+                // since the object passed is already scale decode, offset is 0
+                offset: call_offset(0, false),
+            }),
+            // derive key from the content at the offset
+            FeedObjectMapping::Content { key, offset } => {
+                let key = if let Some(key) = key {
+                    key_namespace(key.as_slice())
+                } else {
+                    // extract the content from the object.
+                    // extract encoded vector length
+                    let mut start = offset as usize;
+
+                    // since this wont be called in runtime, decode Compact<u32> by a new
+                    // allocation of just the encoded bytes of length.
+                    let data = match object[start] & 3 {
+                        0 => vec![object[start]],
+                        1 => Vec::from(&object[start..start + 2]),
+                        2 => Vec::from(&object[start..start + 4]),
+                        // too big. dont care
+                        _ => return None,
+                    };
+
+                    start += data.len();
+                    let length = match Compact::<u32>::decode(&mut data.as_slice()) {
+                        Ok(length) => length.0 as usize,
+                        _ => return None,
+                    };
+                    content_hasher(&object[start..start + length])
```

Comment:
I believe this should work too:
```rust
                    let length = Compact::<u32>::decode(&mut data.as_slice()).ok()?;
                    content_hasher(&object[offset + Compact(length).encoded_size()..][..length])
```
And this will probably be fairly efficient too (and preferred):
```rust
                    content_hasher(&Vec::<u8>::decode(object[offset..]).ok()?)
```

Reply:
> content_hasher(&Vec::<u8>::decode(object[offset..]).ok()?)

I believe this will need a mut ref to the object. Another allocation maybe?

Reply:
Probably `content_hasher(&Vec::::decode(&mut &object[offset..]).ok()?)`. I didn't try to compile it, consider code snippets to be pseudo-code.